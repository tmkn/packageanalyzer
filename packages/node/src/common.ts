import * as path from "path";
import * as fs from "fs";
import { fileURLToPath } from "url";

import dayjs from "dayjs";

import type { DependencyTypes } from "../../shared/src/reports/Validation.js";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const __dirname = path.dirname(fileURLToPath(import.meta.url));
        // adjust when moving this function to another folder
        const file = path.join(__dirname, "..", "..", "..", "package.json");
        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}
