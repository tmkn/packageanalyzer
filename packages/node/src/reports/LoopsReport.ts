import chalk from "chalk";
import { z } from "zod";

import { LoopUtilities } from "../extensions/utilities/LoopUtilities.js";
import { type IPackage } from "../../../shared/src/package/package.js";
import { getPackageVersionfromString } from "../../../shared/src/visitors/visitor.js";
import {
    AbstractReport,
    type IReportConfig,
    type IReportContext
} from "../../../shared/src/reports/Report.js";
import { BasePackageParameter, TypeParameter } from "../../../shared/src/reports/Validation.js";

const LoopParams = BasePackageParameter.merge(TypeParameter);

export type ILoopParams = z.infer<typeof LoopParams>;

export class LoopsReport extends AbstractReport<ILoopParams> {
    name = `Loop Report`;
    configs: IReportConfig;

    constructor(params: ILoopParams) {
        super(params);

        this.configs = {
            pkg: getPackageVersionfromString(params.package),
            type: params.type
        };
    }

    async report([pkg]: [IPackage], { stdoutFormatter }: IReportContext): Promise<void> {
        const loopPathMap = new LoopUtilities(pkg).loopPathMap;
        const distinctCount: number = [...loopPathMap].reduce((i, [, loops]) => i + loops.size, 0);
        const loopPadding = ("" + distinctCount).length;
        let total = 0;

        stdoutFormatter.writeLine(
            chalk.bold(`${distinctCount} Loop(s) found for ${pkg.fullName}\n`)
        );
        if (distinctCount > 0) {
            stdoutFormatter.writeLine(`Affected Packages:`);
            for (const [pkgName, loopsForPkg] of loopPathMap) {
                const loopCountStr = `${loopsForPkg.size}x`.padStart(5);

                stdoutFormatter.writeLine(`- ${loopCountStr} ${pkgName}`);
            }

            for (const [pkgName, loopsForPkg] of loopPathMap) {
                stdoutFormatter.writeLine(
                    chalk.bgGray(`\n${loopsForPkg.size} Loop(s) found for ${pkgName}`)
                );

                let i = 0;
                for (const loop of loopsForPkg) {
                    const iStr = `${total + i++ + 1}`.padStart(loopPadding);

                    stdoutFormatter.writeLine(`[${iStr}/${distinctCount}] ${loop}`);
                }

                total += loopsForPkg.size;
            }
        }
    }

    override validate(): z.ZodType<ILoopParams> {
        return LoopParams;
    }
}
