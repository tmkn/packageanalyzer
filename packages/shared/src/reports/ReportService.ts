import chalk from "chalk";

import { type IPackage } from "../package/package.js";
import { Formatter, type IFormatter } from "../utils/formatter.js";
import { Visitor } from "../visitors/visitor.js";
import { type GenericReport, type IReportConfig, isReportConfigArray } from "./Report.js";
import type { IHost } from "../host/IHost.js";

export type ReportServiceConfig = {
    reports: GenericReport[];
};

export class ReportService {
    private readonly _stdoutFormatter: IFormatter;
    private readonly _stderrFormatter: IFormatter;

    constructor(
        private readonly _config: ReportServiceConfig,
        private readonly _host: IHost
    ) {
        this._stdoutFormatter = new Formatter(this._host.getStdoutWriter());
        this._stderrFormatter = new Formatter(this._host.getStderrWriter());
    }

    async process(): Promise<number | void> {
        let exitCode: number = 0;

        try {
            exitCode = await this._report(this._config);
        } catch (e: any) {
            this._stderrFormatter.writeLine(e?.toString());
            console.error(e?.toString());

            exitCode = 1;
        } finally {
            await Promise.all([
                this._host.getStdoutWriter().flush(),
                this._host.getStderrWriter().flush()
            ]);
        }

        return exitCode;
    }

    private async _report({ reports }: ReportServiceConfig): Promise<number> {
        for (const report of reports) {
            const packages: IPackage[] = await this._getPackages(report);

            if (reports.length > 1)
                this._stdoutFormatter.writeLine(chalk.underline.bgBlue(`Report: ${report.name}`));

            await report.report(packages, {
                stdoutFormatter: this._stdoutFormatter,
                stderrFormatter: this._stderrFormatter
            });
            this._stdoutFormatter.writeLine(``);
        }

        return Math.max(...reports.map(report => report.exitCode));
    }

    private async _getPackage(entry: IReportConfig, report: GenericReport): Promise<IPackage> {
        const visitor = new Visitor(
            entry.pkg,
            report.provider ?? this._host.getDefaultProvider(),
            this._host.getLogger(),
            entry.attachments ?? {},
            entry.depth
        );

        return visitor.visit(entry.type);
    }

    private async _getPackages(report: GenericReport): Promise<IPackage[]> {
        const entries: Array<IReportConfig> = isReportConfigArray(report.configs)
            ? report.configs
            : [report.configs];
        const packageArgs: IPackage[] = [];

        for (const entry of entries) {
            packageArgs.push(await this._getPackage(entry, report));
        }

        return packageArgs;
    }
}
