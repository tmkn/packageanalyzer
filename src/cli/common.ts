import * as path from "path";
import * as fs from "fs";

import * as dayjs from "dayjs";
import * as chalk from "chalk";

import { PackageAnalytics, VersionSummary, GroupedLicenseSummary } from "../analyzers/package";
import { DependencyTypes } from "../visitors/visitor";

export const defaultDependencyType: DependencyTypes = "dependencies";

export function getVersion(): string {
    try {
        const file = path.join(__dirname, "./../../package.json");

        return JSON.parse(fs.readFileSync(file, "utf8")).version;
    } catch (e) {
        return "version parse error!";
    }
}

export function daysAgo(date: string | number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}

export function printStatistics(pa: PackageAnalytics): void {
    const padding = 40;
    const paddingLeft = 4;

    console.log(chalk.bold(`Statistics for ${pa.fullName}\n`));

    printPublished(pa, padding);
    printOldest(pa.oldest, padding);
    printNewest(pa.newest, padding);
    console.log(`${`Direct dependencies:`.padEnd(padding)}${pa.directDependencyCount}`);
    console.log(`${`Transitive dependencies:`.padEnd(padding)}${pa.transitiveDependenciesCount}`);
    printDistinctDependencies(pa.distinctByNameCount, pa.distinctByVersionCount, paddingLeft);
    printMostReferred(pa.mostReferred, padding);
    printMostDependencies(pa.mostDependencies, padding);
    printMostVersion(pa.mostVersions, padding, paddingLeft);
    printLoops(pa, padding, paddingLeft);
    printLicenseInfo(pa.licensesByGroup, paddingLeft);
}

function printNewest(newest: PackageAnalytics | undefined, padding: number): void {
    if (newest && newest.published) {
        console.log(
            `${`Newest package:`.padEnd(padding)}${
                newest.fullName
            } - ${newest.published.toUTCString()} ${daysAgo(newest.published)}`
        );
        console.log(`${`Newest package path:`.padEnd(padding)}${newest.pathString}`);
    }
}

function printOldest(oldest: PackageAnalytics | undefined, padding: number): void {
    if (oldest && oldest.published) {
        console.log(
            `${`Oldest package:`.padEnd(padding)}${
                oldest.fullName
            } - ${oldest.published.toUTCString()} ${daysAgo(oldest.published)}`
        );
        console.log(`${`Oldest package path:`.padEnd(padding)}${oldest.pathString}`);
    }
}

function printPublished(pa: PackageAnalytics, padding: number): void {
    if (!pa.published) return;

    console.log(
        `${`Published:`.padEnd(padding)}${pa.published.toUTCString()} ${daysAgo(pa.published)}`
    );
}

function printDistinctDependencies(
    byName: number,
    byNameAndVersion: number,
    paddingLeft: number
): void {
    console.log(`Distinct dependencies:`);
    console.log(`${``.padStart(paddingLeft)}${byName}: distinct name`);
    console.log(`${``.padStart(paddingLeft)}${byNameAndVersion}: distinct name and version`);
}

function printLoops(pa: PackageAnalytics, padding: number, paddingLeft: number): void {
    const { loops, loopPathMap, distinctLoopCount } = pa;

    console.log(`${`Loops:`.padEnd(padding)}${loops.length} (${distinctLoopCount} distinct)`);

    if (distinctLoopCount > 0) {
        const [first] = loops.map(l => l.pathString).sort();

        console.log(`    affected Packages: [${[...loopPathMap.keys()].join(", ")}]`);
        console.log(`    e.g. ${first}`);

        if (loops.length > 1)
            console.log(`${``.padStart(paddingLeft)}+${loops.length - 1} additional loops`);
    }
}

function printMostDependencies(pa: PackageAnalytics, padding: number): void {
    console.log(
        `${`Most direct dependencies:`.padEnd(padding)}"${pa.name}": ${pa.directDependencyCount}`
    );
}

function printMostReferred(arg: [string, number], padding: number): void {
    console.log(`${`Most referred package:`.padEnd(padding)}"${arg[0]}": ${arg[1]}x`);
}

function printMostVersion(mostVerions: VersionSummary, padding: number, paddingLeft: number): void {
    console.log(`${`Package(s) with most versions:`.padEnd(padding)}`);

    for (const [name, versions] of mostVerions) {
        console.log(`${``.padStart(paddingLeft)}${name} [${[...versions].join(", ")}]`);
    }
}

function printLicenseInfo(groupedLicenses: GroupedLicenseSummary, paddingLeft: number): void {
    console.log(`Licenses:`);
    for (const { license, names } of groupedLicenses) {
        const threshold = 4;
        const samples: string[] = names.slice(0, threshold);

        if (names.length >= threshold)
            console.log(
                `${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}, +${
                    names.length - threshold
                } more]`
            );
        else console.log(`${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}]`);
    }
}
