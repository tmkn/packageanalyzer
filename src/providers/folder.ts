import * as path from "path";
import * as fs from "fs";
import * as semver from "semver";

import type { IPackageJson, IPackageMetadata, IUnpublishedPackageMetadata } from "../npm.js";
import { AbstractPackageProvider, type IPackageJsonProvider } from "./provider.js";

// todo replace in unit tests with FolderPackageProvider
// aka make FolderPackageProvider the default for unit tests

//gathers packages from a node_modules folder
export class FileSystemPackageProvider implements IPackageJsonProvider {
    private readonly _paths: Set<string> = new Set();
    private readonly _cache: Map<string, Map<string, IPackageJson>> = new Map();

    constructor(_folder: string) {
        const matches = this._findPackageJson(_folder);

        this._paths = new Set([...matches].sort((a, b) => a.localeCompare(b)));
        this._load();
    }

    private _findPackageJson(folder: string): string[] {
        try {
            const pkgs: string[] = [];

            const folderEntries = fs
                .readdirSync(folder, "utf8")
                .map(entry => path.join(folder, entry));

            for (const entry of folderEntries) {
                if (fs.statSync(entry).isDirectory()) {
                    pkgs.push(...this._findPackageJson(entry));
                } else if (entry.endsWith(`package.json`)) {
                    pkgs.push(entry);
                }
            }

            return pkgs;
        } catch (e) {
            console.log(e);

            return [];
        }
    }

    private _load(): void {
        const failedPaths: string[] = [];

        for (const pkgPath of this._paths) {
            try {
                const content = fs.readFileSync(pkgPath, "utf8");
                const pkg: IPackageJson = JSON.parse(content);

                this.addPackage(pkg);
            } catch {
                failedPaths.push(pkgPath);

                continue;
            }
        }

        if (failedPaths.length > 0) {
            console.log(`Failed to load ${failedPaths.length} packages`);
        }
    }

    addPackage(pkg: IPackageJson): void {
        const { name, version } = pkg;
        const versions = this._cache.get(name);

        if (typeof versions === "undefined") {
            this._cache.set(name, new Map([[version, pkg]]));
        } else {
            const specificVersion = versions.get(version);

            if (typeof specificVersion === "undefined") {
                versions.set(version, pkg);
            }
        }
    }

    async getPackageJson(name: string, version?: string | undefined): Promise<IPackageJson> {
        const versions = this._cache.get(name);
        let pkg: IPackageJson;

        if (typeof versions === "undefined") {
            throw new Error(`Couldn't find package ${name}`);
        }

        //load latest available version
        if (typeof version === "undefined") {
            const [latestVersion] = [...versions.keys()].slice(-1);

            if (!latestVersion) throw new Error(`Couldn't get latest version for ${name}`);

            const specificVersion = versions.get(latestVersion);

            //should never happen..
            if (typeof specificVersion === "undefined")
                throw new Error(`Error extracting latest package ${name}@${version}`);

            pkg = specificVersion;
        } else {
            const availableVersions: string[] = [...versions.keys()];
            const resolvedVersion = semver.maxSatisfying(availableVersions, version);

            if (resolvedVersion === null) {
                throw new Error(`Couldn't resolve ${version} for ${name}`);
            }

            const specificVersion = versions.get(resolvedVersion);

            if (typeof specificVersion === "undefined")
                throw new Error(`Couldn't find package ${name}@${version}`);

            pkg = specificVersion;
        }

        return pkg;
    }
}

// provides data provided by the dependencydump cli command
export class DumpPackageProvider extends AbstractPackageProvider {
    constructor(private readonly _folder: string) {
        super();

        this._loadData();
    }

    private _findPackageJson(folder: string): Set<string> {
        try {
            const pkgs: string[] = [];

            const folderEntries = fs
                .readdirSync(folder, "utf8")
                .map(entry => path.join(folder, entry));

            for (const entry of folderEntries) {
                if (fs.statSync(entry).isDirectory()) {
                    pkgs.push(...this._findPackageJson(entry));
                } else if (entry.endsWith(`metadata.json`)) {
                    pkgs.push(entry);
                }
            }

            return new Set([...pkgs].sort((a, b) => a.localeCompare(b)));
        } catch (e) {
            console.log(e);

            return new Set();
        }
    }

    private _loadData(): void {
        const paths = this._findPackageJson(this._folder);
        const failedPaths: string[] = [];

        for (const pkgPath of paths) {
            try {
                const content = fs.readFileSync(pkgPath, "utf8");
                const pkg: IPackageMetadata = JSON.parse(content);

                this._cache.set(pkg.name, pkg);
            } catch {
                failedPaths.push(pkgPath);

                continue;
            }
        }

        if (failedPaths.length > 0) {
            console.log(`Failed to load ${failedPaths.length} packages`);
        }
    }

    async getPackageMetadata(
        name: string
    ): Promise<IPackageMetadata | IUnpublishedPackageMetadata | undefined> {
        return this._cache.get(name);
    }
}
