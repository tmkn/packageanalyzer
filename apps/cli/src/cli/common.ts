import { Command } from "clipanion";

import { AbstractReport } from "../../../../packages/shared/src/reports/Report.js";
import { ReportService } from "../../../../packages/shared/src/reports/ReportService.js";
import { Formatter, type IFormatter } from "../../../../packages/shared/src/utils/formatter.js";
import { NodeHost } from "../../../../packages/node/src/host/NodeHost.js";

export abstract class CliCommand<
    T extends AbstractReport<any> | AbstractReport<any>[]
> extends Command {
    abstract getReports(): T | Promise<T>;

    exitCode = 0;

    beforeProcess: ((report: T) => void) | undefined = undefined;

    async execute(): Promise<number | void> {
        const host = new NodeHost(this.context.stdout, this.context.stderr);

        try {
            const reports = await this.getReports();
            const reportService = new ReportService(
                {
                    reports: Array.isArray(reports) ? reports : [reports]
                },
                host
            );

            this.beforeProcess?.(reports);
            this.exitCode = (await reportService.process()) ?? 0;
        } catch (e: unknown) {
            const stderrFormatter: IFormatter = new Formatter(host.getStderrWriter());

            if (e) stderrFormatter.writeLine(e?.toString());
            this.exitCode = 1;
        } finally {
            await Promise.all([host.getStdoutWriter().flush(), host.getStderrWriter().flush()]);
        }

        return this.exitCode;
    }
}
