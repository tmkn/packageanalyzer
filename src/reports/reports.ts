import { IDecorator } from "../extensions/decorators/Decorator";
import { Package } from "../package/package";
import { IPackageVersionProvider } from "../providers/folder";
import { IFormatter } from "../utils/formatter";
import { PackageVersion } from "../visitors/visitor";

export interface IReport<T extends {}> {
    readonly name: string;
    readonly params: T;
    readonly pkg: PackageVersion;

    decorators?: IDecorator<any>[];
    provider?: IPackageVersionProvider;

    report(pkg: Package, formatter: IFormatter): Promise<void>;
}

interface ITestReportParams {}
export class TestReport implements IReport<ITestReportParams> {
    name = `Test Report`;
    pkg: PackageVersion = [``, ``];

    constructor(readonly params: ITestReportParams) {}

    async report(pkg: Package, formatter: IFormatter): Promise<void> {}
}

export class ReportService {
    async process(reports: IReport<any>[]): Promise<void> {
        try {
            for (const report of reports) {
            }
        } catch (e) {}
    }
}
