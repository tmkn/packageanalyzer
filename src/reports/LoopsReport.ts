import * as chalk from "chalk";

import { isValidDependencyType } from "../cli/common";
import { LoopUtilities } from "../extensions/utilities/LoopUtilities";
import { Package } from "../package/package";
import { getPackageVersionfromString, PackageVersion } from "../visitors/visitor";
import { AbstractReport, IReportContext } from "./Report";
import { DependencyTypes } from "./Validation";

export interface ILoopParams {
    package: string;
    type: DependencyTypes;
}

export class LoopsReport extends AbstractReport<ILoopParams> {
    name = `Loop Report`;
    pkg: PackageVersion;

    constructor(override readonly params: ILoopParams) {
        super(params);

        this.pkg = getPackageVersionfromString(params.package);
        this.type = params.type;
    }

    async report(pkg: Package, { stdoutFormatter }: IReportContext): Promise<void> {
        if (!isValidDependencyType(this.type)) {
            throw new Error(
                `Please only specify "dependencies" or "devDependencies" for the --type argument`
            );
        }

        const loopPathMap = new LoopUtilities(pkg).loopPathMap;
        const distinctCount: number = [...loopPathMap].reduce((i, [, loops]) => i + loops.size, 0);
        const loopPadding = ("" + distinctCount).length;
        let total = 0;

        stdoutFormatter.writeLine(
            chalk.bold(`${distinctCount} Loop(s) found for ${pkg.fullName}\n`)
        );
        if (distinctCount > 0) {
            stdoutFormatter.writeLine(`Affected Packages:`);
            for (const [pkgName, loopsForPkg] of loopPathMap) {
                const loopCountStr = `${loopsForPkg.size}x`.padStart(5);

                stdoutFormatter.writeLine(`- ${loopCountStr} ${pkgName}`);
            }

            for (const [pkgName, loopsForPkg] of loopPathMap) {
                stdoutFormatter.writeLine(
                    chalk.bgGray(`\n${loopsForPkg.size} Loop(s) found for ${pkgName}`)
                );

                let i = 0;
                for (const loop of loopsForPkg) {
                    const iStr = `${total + i++ + 1}`.padStart(loopPadding);

                    stdoutFormatter.writeLine(`[${iStr}/${distinctCount}] ${loop}`);
                }

                total += loopsForPkg.size;
            }
        }
    }

    validate = undefined;
}
