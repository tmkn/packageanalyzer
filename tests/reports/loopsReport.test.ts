import * as path from "path";
import { DumpPackageProvider } from "../../src/providers/folder";

import { LoopsReport } from "../../src/reports/LoopsReport";
import { ReportService } from "../../src/reports/ReportService";
import { createMockContext } from "../common";

describe(`LoopsReport Test`, () => {
    const rootPath = path.join("tests", "data", "loops_data");
    let provider: DumpPackageProvider;

    beforeAll(() => {
        provider = new DumpPackageProvider(rootPath);
    });

    test(`works`, async () => {
        const report = new LoopsReport({
            package: `@webassemblyjs/ast@1.9.0`,
            type: `dependencies`
        });

        report.provider = provider;

        const { stdout, stderr } = createMockContext();
        const reportService = new ReportService(
            {
                mode: "distinct",
                reports: [report]
            },
            stdout,
            stderr
        );

        await reportService.process();

        expect(stdout.lines).toMatchSnapshot(`stdout`);
        expect(stderr.lines).toMatchSnapshot(`stderr`);
    });

    test(`Throws on illegal dependency type`, async () => {
        expect.assertions(1);

        try {
            const report = new LoopsReport({
                package: `foo`,
                //@ts-expect-error type needs to be a valid dependency type
                type: `xxx`
            });
        } catch (e) {
            expect(e).toBeInstanceOf(Error);
        }
    });
});
