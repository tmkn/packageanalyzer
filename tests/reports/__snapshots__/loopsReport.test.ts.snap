// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoopsReport Test works: stderr 1`] = `[]`;

exports[`LoopsReport Test works: stdout 1`] = `
[
  "4 Loop(s) found for @webassemblyjs/ast@1.9.0",
  "",
  "Affected Packages:",
  "-    3x @webassemblyjs/ast",
  "-    1x @webassemblyjs/wast-parser",
  "",
  "3 Loop(s) found for @webassemblyjs/ast",
  "[1/4] @webassemblyjs/ast@1.9.0 → @webassemblyjs/helper-module-context@1.9.0 → @webassemblyjs/ast@1.9.0",
  "[2/4] @webassemblyjs/ast@1.9.0 → @webassemblyjs/wast-parser@1.9.0 → @webassemblyjs/ast@1.9.0",
  "[3/4] @webassemblyjs/ast@1.9.0 → @webassemblyjs/wast-parser@1.9.0 → @webassemblyjs/helper-code-frame@1.9.0 → @webassemblyjs/wast-printer@1.9.0 → @webassemblyjs/ast@1.9.0",
  "",
  "1 Loop(s) found for @webassemblyjs/wast-parser",
  "[4/4] @webassemblyjs/wast-parser@1.9.0 → @webassemblyjs/helper-code-frame@1.9.0 → @webassemblyjs/wast-printer@1.9.0 → @webassemblyjs/wast-parser@1.9.0",
  "",
  "",
]
`;
